server/src/main/java/org/candlepin/controller/Entitler.java

```diff
@@ -376,47 +382,67 @@ private DeveloperProducts getDeveloperPoolProducts(Owner owner, String sku) {
      *         from the adapter.
      */
     private DeveloperProducts getDevProductMap(Owner owner, String sku) {
-
         Collection<? extends ProductInfo> productsByIds = this.productAdapter
             .getProductsByIds(owner.getKey(), Arrays.asList(sku));
 
-        Collection<String> devProductIds = new ArrayList<>();
-        if (productsByIds.iterator().hasNext()) {
+        Map<String, Product> devProductMap = new HashMap<>();
+
+        if (productsByIds != null && !productsByIds.isEmpty()) {
+            // We're apparently only interested in the first product returned for the given sku
             ProductInfo devProduct = productsByIds.iterator().next();
-            devProductIds.add(devProduct.getId());
 
-            Collection<? extends ProductInfo> provided = devProduct.getProvidedProducts();
+            // Collect the dev product IDs from the potential tree of products we received
+            List<String> devProductIds = new ArrayList<>();
+            this.collectDevProductIds(devProductIds, devProduct);
 
-            if (provided != null) {
-                provided.forEach(product -> {
-                    if (product != null) {
-                        devProductIds.add(product.getId());
-                    }
-                });
-            }
-        }
+            // Do a refresh so we're all up to date here
+            log.debug("Importing products for dev pool resolution...");
 
-        log.debug("Importing products for dev pool resolution...");
-        ImportedEntityCompiler compiler = new ImportedEntityCompiler();
+            RefreshWorker refresher = this.refreshWorkerProvider.get();
+            refresher.addProducts(this.productAdapter.getProductsByIds(owner.getKey(), devProductIds));
+            RefreshResult refreshResult = refresher.execute(owner);
 
-        compiler.addProducts(this.productAdapter.getProductsByIds(owner.getKey(), devProductIds));
+            // Step through the items we refreshed and add the resulting products to our map
+            for (String pid : refresher.getProducts().keySet()) {
+                Product product = refreshResult.getProduct(pid);
+                if (product != null) {
+                    devProductMap.put(product.getId(), product);
```

Comment:
You mentioned yesterday that because of how the RefreshWorker works, we're adding all the org's products to the `devProductMap`. Can we not do some filtering here, and only add the refreshed products that are part of `sku` (and its hierarchy below it)?

Reply:
I don't fully understand the question.

This block is what is doing that filtering. Previously we were relying on the poorly named .getImportedProducts to grab everything, but with the new system, we need to be a little more selective to get the correct result.

Part of the issue here is how DeveloperProduct works, and still expects to be working with a flat map. I feel there are going to be many places like this where we have to make a decision to either update a larger area of code to be more n-tier aware, or have patches like this which allow those places to continue working as they do.

Reply:
Nevermind, I misread the code!